libgnucash/engine/Account.cpp

```diff
@@ -3578,27 +3559,16 @@ xaccAccountGetProjectedMinimumBalance (const Account *acc)
 static gnc_numeric
 GetBalanceAsOfDate (Account *acc, time64 date, std::function<gnc_numeric(Split*)> split_to_numeric)
 {
-    /* Ideally this could use xaccAccountForEachSplit, but
-     * it doesn't exist yet and I'm uncertain of exactly how
-     * it would work at this time, since it differs from
-     * xaccAccountForEachTransaction by using gpointer return
-     * values rather than gints.
-     */
-    Split *latest = nullptr;
-
     g_return_val_if_fail(GNC_IS_ACCOUNT(acc), gnc_numeric_zero());
 
     xaccAccountSortSplits (acc, TRUE); /* just in case, normally a noop */
     xaccAccountRecomputeBalance (acc); /* just in case, normally a noop */
 
-    for (GList *lp = GET_PRIVATE(acc)->splits; lp; lp = lp->next)
-    {
-        if (xaccTransGetDate (xaccSplitGetParent ((Split *)lp->data)) >= date)
-            break;
-        latest = (Split *)lp->data;
-    }
+    auto is_before_date = [date](auto s) -> bool
+    { return xaccTransGetDate(xaccSplitGetParent(s)) < date; };
 
-    return latest ? split_to_numeric (latest) : gnc_numeric_zero();
+    auto latest_split{gnc_account_find_split (acc, is_before_date, true)};
```

Comment:
This would be an ideal use for `std::upper_bound`

Reply:
... or not. IRL most GetBalanceAsOfDate's dates would be towards the later end of the splits vector...